GVFS/GVFS.Common/Database/Placeholders.cs

```diff
@@ -0,0 +1,199 @@
+﻿using Microsoft.Data.Sqlite;
+using System;
+using System.Collections.Generic;
+using System.Data;
+
+namespace GVFS.Common.Database
+{
+    public class Placeholders : IPlaceholderDatabase
+    {
+        private GVFSDatabase database;
+
+        public Placeholders(GVFSDatabase database)
+        {
+            this.database = database;
+        }
+
+        public static void CreateTable(SqliteCommand command)

```

Comment:
Related to comment on having `CreateTable` take a `SqliteCommand`, I'm having a bit of trouble understanding the relationship between `Placeholders` and `GVFSDatabase`.

- Why is `CreateTable` `static`?  If `Placeholders` has a `GVFSDatabase` member, could it just request a connection from the pool (like `Count` does)?
- If `Placeholders` is tied this closely to `GVFSDatabase`, should it be a nested class of `GVFSDatabase`?
- What are the pros/cons of making `Placeholders` is own class vs accessing placeholder data via `GVFSDatabase`?

Reply:
So a little information on how I arrived here.  I originally had this code in the `GVFSDatabase` class because it is the class that is initializing the database and creating all the tables for the database but when creating the `Placeholders` class, I felt this needed to be a part of this class since this is the class that will be using the fields of the table and for easy reference to what the fields of the table would be I put them here.

>Why is CreateTable static? If Placeholders has a GVFSDatabase member, could it just request a connection from the pool (like Count does)?

Because `Placeholders` needs a `GVFSDatabase` but the `GVFSDatabase` needs to create the tables of the database in it's constructor there is kind of a chicken/egg problem here if it is not `static`.  I supposed we could have the `Placeholders` try to create the table in its constructor but I felt that was the responsibility of the `GVFSDatabase` class to make sure the database is initialized and ready to use.

>If Placeholders is tied this closely to GVFSDatabase, should it be a nested class of GVFSDatabase?

To me this is a separate issue that I can see either way.  I am making some changes so that I can unit test this class that will decouple this class from the `GVFSDatabase` class but it could still be nested because it is part of the `GVFSDatabase` class in a way.  I would still want to keep the ability to unit test it though so it would need to be `public` which would be an argument for not have it be nested.

>What are the pros/cons of making Placeholders is own class vs accessing placeholder data via GVFSDatabase?

The biggest con for putting everything in the `GVFSDatabase` is that it would grow and become harder and harder to maintain if we add modified paths, repo metadata, other tables, etc.  I guess the pro would be all the database stuff would be in one class?  I'm one for smaller classes that have specific responsibilities vs one that can do everything.

Reply:
Thanks for all the details here.  What would you think of renaming `Placeholders` to `PlaceholdersTable`?

I think if the name of the class reflects that this class is specifically for interacting with a table in the database it might add some clarity to how/why responsibilities are split between these classes.

Reply:
Actually, maybe put that suggestion on hold.  

I think there might be a few classes/interfaces to consider when it comes to naming `Placeholders`:

- `IPlaceholderDatabase` 
- `Placeholders`
- `GVFSDatabase`

If `IPlaceholderDatabase` doesn't have any "database" specific methods, maybe something like `IPlaceholderCollection` would be a more descriptive name?

And now I am torn on if the `Placeholders` name should have `Table` in it.  Do you see this class representing a table in the `GVFSDatabase`?

Reply:
>If IPlaceholderDatabase doesn't have any "database" specific methods, maybe something like IPlaceholderCollection would be a more descriptive name?

Yeah that sounds like a good change.

>Do you see this class representing a table in the GVFSDatabase?

Right now it is but it might not be the only one in the future.  For example if we add the modified paths to the database as another table and we have a method on a ModifiedPaths Table class that is Add but that method will also delete from the Placeholder Table then things are not as straight forward.  At that point though we may need another class that is more like the business logic that adds to the modified paths table and deletes from the placeholder table using a transaction.  So I think for this PR we can rename it to have `Table` in it.